apps/settings/js/modules/addon_manager.js

```diff
@@ -0,0 +1,293 @@
+
+      if (needsReboot) {
+        return 'reboot';
+      }
+
+      var needsRestart = customizedApps.some(function(customizedApp) {
+        return customizedApp.customizations.some((c) => c.scripts.length > 0);
+      });
+
+      if (needsRestart) {
+        return 'restart';
+      }
+
+      return 'disabled';
+    });
+  }
+
+  function canDelete(addon) {
+    return isAddon(addon) && addon.removable === true;
+  }
+
+  function deleteAddon(addon) {
+    if (!isAddon(addon)) {
+      return Promise.reject('not an addon');
+    }
+
+    if (!addon.removable) {
+      return Promise.reject('addon is not deletable');
+    }
+
+    return new Promise(function(resolve, reject) {
+      var request = mozApps.mgmt.uninstall(addon);
+      request.onsuccess = function() {
+        resolve();
+      };
+      request.onerror = function() {
+        reject();
+      };
+    });
+  }
+
+  function shareAddon(addon) {
+    if (!isAddon(addon)) {
+      return Promise.reject('not an addon');
+    }
+
+    return new Promise(function(resolve, reject) {
+      var activity = new MozActivity({
+        name: 'share',
+        data: {
+          type: 'app',
+          app: addon.manifestURL
+        }
+      });
+
+      activity.onsuccess = function() {
+        resolve();
+      };
+      activity.onerror = function() {
+        reject((activity.error && activity.error.name) || 'activity error');
+      };
+    });
+  }
+
+  // Return an array of the apps that are targeted by any of this
+  // addon's customizations. Note, however, that addons can also target
+  // arbitrary web pages and hosted apps. This test only applies to
+  // installed packaged apps and so may be of limited utility.
+  function getAddonTargets(addon) {
+    return getCustomizedApps(addon).then(function(customizedApps) {
+      return customizedApps.map((x) => x.app);
+    });
+  }
+
+  // Addons can only apply to apps with the same or lower privilege
+  function privilegeCheck(addon, app) {
+    var addonManifest = getManifest(addon);
+    var appManifest = getManifest(app);
+    return addonManifest.type === appManifest.type ||
+      addonManifest.type === 'certified' ||
+      (addonManifest.type === 'privileged' && appManifest.type !== 'certified');
+  }
+
+  // This internal utility function returns an array of objects describing
+  // the aps that are affected by this addon. Each object in the array has
+  // an 'app' property that holds the app object and an 'customizations'
+  // property that is an array of customizations (from the addon manifest).
+  // We implement getAddonTargets() using this information and also derive
+  // the 'reboot' and 'restart' return values of disableAddon() from it.
+  //
+  // We need to keep this code in sync with the matching gecko
+  // code in dom/apps/UserCustomizations.jsm so that our list of
+  // targeted apps actually matches the apps that Gecko injects
+  // the addon into.
+  //
+  function getCustomizedApps(addon) {
```

Comment:
I have 71 apps on my flame, and calls to getCustomizedApps() take about 50ms. I was very surprised that it was that slow. Fabrice does say that accessing the properties of App objects may go through a proxy so that might be part of the slowness.

Even though this is slow, it is done asynchronously, so it should not slow down the responsiveness of the panel transition, right?  

I'd really prefer not to implement another layer of caching here. It seems premature, at least for now when addons are a developer-only feature and we don't have any actual performance complaint.

I'll add a note in the comment that it is a relatively expensive function and mention the idea that we consider a cache here.